marimo/_runtime/executor.py

```diff
@@ -0,0 +1,389 @@
+                    or is_unclonable_type(glbls[ref])
+                ):
+                    lcls[ref] = glbls[ref]
+                elif isinstance(glbls[ref], ShallowCopy):
+                    lcls[ref] = shallow_copy(glbls[ref])
+                else:
+                    try:
+                        lcls[ref] = deepcopy(glbls[ref])
+                    except TypeError as e:
+                        raise CloneError(
+                            f"Could not clone reference `{ref}` of type "
+                            f"{getattr(glbls[ref], '__module__', '<module>')}."
+                            f"{glbls[ref].__class__.__name__}"
+                            " try wrapping the object in a `zero_copy`"
+                            "call. If this is a common object type, consider "
+                            "making an issue on the marimo GitHub "
+                            "repository to never deepcopy."
+                        ) from e
+            elif ref not in glbls["__builtins__"]:
+                if ref in cell.defs:
+                    raise MarimoNameError(
+                        f"name `{ref}` is referenced before definition."
+                    )
+                raise MarimoMissingRefError(ref)
+
+        # NOTE: Execution expects the globals dictionary by memory reference,
+        # so we need to clear it and update it with the sanitized locals,
+        # returning a backup of the original globals for later restoration.
+        # This must be performed at the end of the function to ensure valid
+        # state in case of failure.
+        backup = {**glbls}
+        glbls.clear()
+        glbls.update(lcls)
+        return backup
+
+    @staticmethod
+    def update_outputs(
+        cell: CellImpl, glbls: dict[str, Any], backup: dict[str, Any]
+    ) -> None:
+        # NOTE: After execution, restore global state and update outputs.
+        lcls = {**glbls}
+        glbls.clear()
+        glbls.update(backup)
+
+        defs = cell.defs
+        for df in defs:
+            if df in lcls:
+                # Overwrite will delete the reference.
+                # Weak copy holds on with references.
+                glbls[df] = lcls[df]
+            elif df in glbls:
+                del glbls[df]
+
+        for df in backup:
+            if df.startswith(f"_cell_{cell.cell_id}_"):
+                del glbls[df]
+
+        from marimo._plugins.ui._core.ui_element import UIElement
+
+        # Private UI elements should still be in memory.
+        # Or a required ref.
+        required = set()
+        for variable in cell.variable_data.values():
+            required.update(variable.required_refs)
+
+        for df in lcls:
+            if df.startswith(f"_cell_{cell.cell_id}_"):
+                if df in required or isinstance(lcls[df], UIElement):
+                    glbls[df] = lcls[df]
+
+
+def build_ref_predicate(
+    glbls: dict[str, Any],
+) -> Callable[[Name, VariableData], bool]:
```

Comment:
Can you add one sentence somewhere that describes what the returned callable does?

Something like: Returns a callable that takes a variable name and its associated data, and returns `True`  if ...